### PR TITLE
Call AcquireTokenInteractively instead of SignInInteractively

### LINKED ISSUE #3437: [Bug] Call AcquireTokenInteractively instead of SignInInteractively
Tracking item to investigate and test all scenarios before doing the switch from SignInInteractively -> AcquireTokenInteractively 

If you knew the account is in WAM and am able to rehydrate the account, please do call AcquireTokenInteractively instead of SignInInteractively. We don’t want to show any unnecessary prompt to the user. 

**Background :** 

When the account is not the default OS Account or when no login_hint is passed then we call the [SIgnInInteractivelyAsync API](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzureAD%2Fmicrosoft-authentication-library-for-dotnet%2Fblob%2Fmain%2Fsrc%2Fclient%2FMicrosoft.Identity.Client.Broker%2FRuntimeBroker.cs%23L83&data=05%7C01%7CGladwin.Johnson%40microsoft.com%7Ca54e71ad965146040edc08da5ab6933a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637922037460050967%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=S%2BZCdFBUdOP%2FVUQdouL%2FYwCjmdRAAp84OzSWzvKdTwU%3D&reserved=0), Now you get prompted to let windows remember the account for further sign-ins. Choosing to remember.  WAM goes through this registration process.

**Expected :** With that message “Windows will remember the account”, I’d expect the UI is not shown when I pass the login_hint for subsequent calls. But instead the UI is always thrown requesting to enter password. 

**Suggested Fix :** Call AcquireTokenInteractively instead of SignInInteractively

### PR DESCRIPTION
**Pull request by gladjohn**
Fixes #3437

**Changes proposed in this request**
If we know the account is in WAM and am able to rehydrate the account, we will now call AcquireTokenInteractively instead of SignInInteractively. We don’t want to show any unnecessary prompt to the user.

**Testing**
Dev App in different OSs

**Performance impact**
none, better user experience

### CODE REVIEW

**Comment by gladjohn** [comment]
this needs to be updated based on my other PR (0.9.0 update)

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 1]
```diff
@@ -77,26 +77,58 @@ internal class RuntimeBroker : IBroker
...
+                                _logger.Verbose("[WamBroker] Successfully retrieved token.");
+                            }
+                            else
+                            {
+                                _logger.Error($"[WamBroker] Could not login interactively. {result.Error}");
+                                WamAdapters.ThrowExceptionFromWamError(result, authenticationRequestParameters, _logger);
+                            }
+                        }
                     }
+
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:130]: nit: extra line

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 2]
```diff
@@ -77,26 +77,58 @@ internal class RuntimeBroker : IBroker
...
-                        _logger.Error($"[WamBroker] Could not login interactively. {result.Error}");
-                        WamAdapters.ThrowExceptionFromWamError(result, authenticationRequestParameters, _logger);
+                        using (var result = await core.AcquireTokenInteractivelyAsync(
+                        _parentHandle,
+                        authParams,
+                        authenticationRequestParameters.CorrelationId.ToString("D"),
+                        wamAccount,
+                        cancellationToken).ConfigureAwait(false))
+                        {
+                            if (result.IsSuccess)
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:118]: Line #118 to #127 are same as Line 92-101. Optimize please

**Review by SameerK-MSFT — APPROVED** [review]

**Review by neha-bhargava — APPROVED** [review]

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 3]
```diff
@@ -77,25 +78,38 @@ internal class RuntimeBroker : IBroker
...
                 _logger.Verbose("[WamBroker] AcquireTokenInteractive - login hint provided? " + string.IsNullOrEmpty(loginHint));
 
-                using (var result = await core.SignInInteractivelyAsync(
-                    _parentHandle,
-                    authParams,
-                    authenticationRequestParameters.CorrelationId.ToString("D"),
-                    loginHint,
-                    cancellationToken).ConfigureAwait(false))
+                //if login hint is not provided or if account is null
+                if (string.IsNullOrEmpty(authenticationRequestParameters.LoginHint))
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:84]: If `LoginHint` is provided, then `Account` on Line 101 will never be null?

File: src/client/Microsoft.Identity.Client.Broker/WamAdapters.cs [hunk 4]
```diff
@@ -182,6 +182,26 @@ private static void AddPopParams(AuthenticationRequestParameters authenticationR
             }
         }
 
+        public static MsalTokenResponse AcquireTokenFromBroker(NativeInterop.AuthResult authResult,
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client.Broker/WamAdapters.cs:185]: Rename to something like HandleResponse, ProcessResponse, ParseResponse.
nit: put first param on new line.

**Review by pmaytak — APPROVED** [review]

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs (hunk 3 above)
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:84]: shouldn't this use "loginHint" from line 80?

File: src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs [hunk 5]
```diff
@@ -77,25 +78,38 @@ internal class RuntimeBroker : IBroker
...
+                        msalTokenResponse = WamAdapters.AcquireTokenFromBroker(result, authenticationRequestParameters, _logger, errorMessage);
+
                     }
-                    else
+                }
+                //if an account is already signed in
+                else
+                {
+                    using (var wamAccount = await core.ReadAccountByIdAsync(
+                    authenticationRequestParameters.Account.HomeAccountId.ObjectId,
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client.Broker/RuntimeBroker.cs:101]: I think this logic needs to be cleaned up a bit. It should more like: 

1. If an Account object is present and if that object has an ID and Object Id THEN do `AcquireTokenInteractivelyAsync`
2. If a loginHint is present (as defined at line 80), then do SignInInteractivelyAsync

Remember that we don't force app developers to use our Account objects. THey only have an interface called IAccount (in hinghsight, this was a mistake). I have seen a LOT of ppl creating their own Account objects which ignore the ID.

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
Account obj might be corrupted, you need to do more defensive programming.

**Comment by gladjohn** [comment]
> Account obj might be corrupted, you need to do more defensive programming.

handled now

**Review by pmaytak — APPROVED** [review]

File: src/client/Microsoft.Identity.Client.Broker/WamAdapters.cs [hunk 6]
```diff
@@ -182,6 +182,27 @@ private static void AddPopParams(AuthenticationRequestParameters authenticationR
...
+        public static MsalTokenResponse HandleResponse(
+                NativeInterop.AuthResult authResult,
+                AuthenticationRequestParameters authenticationRequestParameters,
+                ILoggerAdapter logger, string errorMessage = null)
+        {
+            MsalTokenResponse msalTokenResponse = null;
+
+            if (authResult.IsSuccess)
+            {
+                msalTokenResponse = WamAdapters.ParseRuntimeResponse(authResult, authenticationRequestParameters, logger);
```
[the remark this answers is not in the document]
Comment (gladjohn) [reply src/client/Microsoft.Identity.Client.Broker/WamAdapters.cs:194]: if parsing fails then we throw `MsalRuntimeException`

**Pull request merged**